python/lib/core/dmod/core/common/collections/cache.py

```diff
@@ -0,0 +1,1015 @@
+"""
+@TODO: Put a module wide description here
+"""
+from __future__ import annotations
+
+import asyncio
+import inspect
+import typing
+import math
+
+from datetime import datetime
+
+try:
+    from typing import ParamSpec
+    from typing import Concatenate
+except ImportError:
+    from typing_extensions import ParamSpec
+    from typing_extensions import Concatenate
+
+from .constants import EntryType
+
+from ...events import Event
+from ...events import EventRouter
+
+ON_ADDITION_KEY = "on_addition"
+ON_REMOVAL_KEY = "on_removal"
+ON_ACCESS_KEY = "on_access"
+ON_UPDATE_KEY = "on_update"
+
+HashableType = typing.TypeVar("HashableType", bound=typing.Union[typing.Hashable, typing.Mapping, typing.Sequence[typing.Hashable]])
+
+
+_VARIABLE_PARAMETERS = ParamSpec("_VARIABLE_PARAMETERS")
+"""Represents *args and **kwargs"""
+
+
+def hash_hashable_map_sequence(value: HashableType) -> int:
+    """
+    Generate the hash or hash of a collection of objects
+
+    Args:
+        value: The value to be hashed
+
+    Returns:
+        The result of the hashing operation
+    """
+    if not isinstance(value, (str, bytes)) and isinstance(value, typing.Sequence):
+        return hash(
+            (
+                hash_hashable_map_sequence(item) for item in value
+            )
+        )
+    elif isinstance(value, typing.Mapping):
+        return hash(
+            (
+                (
+                    key if isinstance(key, typing.Hashable) else hash_hashable_map_sequence(key),
+                    value if isinstance(value, typing.Hashable) else hash_hashable_map_sequence(value)
+                )
+                for key, value in sorted(value.items())
```

Comment:
Just addressed this - while something like objects as keys shouldn't ever happen, I beefed up the sorting logic to key on the hash of the key or just a quick string representation of it. The significance isn't that it's a genius ordering, just that it is consistent on things that happen to be equal (it can be out of order if they aren't equal - that doesn't matter).